File: 0039-combination-sum/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.backtrack(candidates,0,target,[])
        return self.result

    def backtrack(self, candidates, k, left, path):
        if left == 0:
            self.result.append(path[:])
            return
        if k == len(candidates):
            return
        for i in range(left //candidates[k] +1):
            for j in range(i):
                path.append(candidates[k])
            self.backtrack(candidates, k+1, left - candidates[k] * i, path)
            for j in range(i):
                path.pop()

File: 0039-combination-sum/cpp/Solution.cpp
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> path;
        backtrack(candidates, 0, target, path);
        return result;
    }
    void backtrack(vector<int> candidates, int k, int left, vector<int> &path) {
        if (left == 0) {
            result.push_back(path);
            return;
        }
        if (k == candidates.size()) {
            return;
        }
        for (int i = 0; i <= left / candidates[k]; ++i) {
            for (int j = 0; j < i; ++j) {
                path.push_back(candidates[k]);
            }
            backtrack(candidates, k + 1, left - i * candidates[k], path);
            for (int j = 0; j < i; ++j) {
                path.pop_back();
            }
        }
    }
};